runtime/parachains/src/paras.rs

```diff
@@ -791,32 +1107,249 @@ impl<T: Config> Pallet<T> {
+				},
+				PvfCheckCause::Upgrade { id, relay_parent_number } => {
+					weight +=
+						Self::proceed_with_upgrade(*id, code_hash, now, *relay_parent_number, cfg);
+				},
+			}
+		}
+		weight
+	}
+
+	fn proceed_with_onboarding(id: ParaId, sessions_observed: SessionIndex) -> Weight {
+		let weight = T::DbWeight::get().reads_writes(2, 1);
+
+		// we should onboard only after `SESSION_DELAY` sessions but we should take
+		// into account the number of sessions the PVF pre-checking occupied.
+		//
+		// we cannot onboard at the current session, so it must be at least one
+		// session ahead.
+		let onboard_at: SessionIndex = shared::Pallet::<T>::session_index() +
+			cmp::max(shared::SESSION_DELAY.saturating_sub(sessions_observed), 1);
+
+		ActionsQueue::<T>::mutate(onboard_at, |v| {
+			if let Err(i) = v.binary_search(&id) {
+				v.insert(i, id);
+			}
+		});
+
+		weight
+	}
+
+	fn proceed_with_upgrade(
+		id: ParaId,
+		code_hash: &ValidationCodeHash,
+		now: T::BlockNumber,
+		relay_parent_number: T::BlockNumber,
+		cfg: &configuration::HostConfiguration<T::BlockNumber>,
+	) -> Weight {
+		let mut weight = 0;
+
+		// Compute the relay-chain block number starting at which the code upgrade is ready to be
+		// applied.
+		//
+		// The first parablock that has a relay-parent higher or at the same height of `expected_at`
+		// will trigger the code upgrade. The parablock that comes after that will be validated
+		// against the new validation code.
+		//
+		// Here we are trying to choose the block number that will have `validation_upgrade_delay`
+		// blocks from the relay-parent of the block that schedule code upgrade but no less than
+		// `minimum_validation_upgrade_delay`. We want this delay out of caution so that when
+		// the last vote for pre-checking comes the parachain will have some time until the upgrade
+		// finally takes place.
+		let expected_at = cmp::max(
+			relay_parent_number + cfg.validation_upgrade_delay,
+			now + cfg.minimum_validation_upgrade_delay,
+		);
+
+		weight += T::DbWeight::get().reads_writes(1, 4);
+		FutureCodeUpgrades::<T>::insert(&id, expected_at);
+
+		<Self as Store>::UpcomingUpgrades::mutate(|upcoming_upgrades| {
+			let insert_idx = upcoming_upgrades
+				.binary_search_by_key(&expected_at, |&(_, b)| b)
+				.unwrap_or_else(|idx| idx);
+			upcoming_upgrades.insert(insert_idx, (id, expected_at));
+		});
+
+		let expected_at = expected_at.saturated_into();
+		let log = ConsensusLog::ParaScheduleUpgradeCode(id, *code_hash, expected_at);
+		<frame_system::Pallet<T>>::deposit_log(log.into());
+
+		weight
+	}
+
+	fn enact_pvf_rejected(
+		code_hash: &ValidationCodeHash,
+		causes: Vec<PvfCheckCause<T::BlockNumber>>,
+	) -> Weight {
+		let mut weight = T::DbWeight::get().writes(1);
+
+		for cause in causes {
+			// Whenever PVF pre-checking is started or a new cause is added to it, the RC is bumped.
+			// Now we need to unbump it.
+			weight += Self::decrease_code_ref(code_hash);
```

Comment:
Yes. 

The idea is that paras retains the code whenever it encounters that code. That would be the code added by either onboarding or upgrading. Each of these processes will bump the RC, since they will each be initiated by a different para. If PVF pre-checking failed though we need to get rid of the code, hence we decrease the RC here.

Also there is an invariant: we assume that after iterating over all causes the RC should reach 0. Why? Because when we just started PVF pre-checking, we checked that there is no such code [^1] (If there were, then the pre-checking wouldn't have started and the code was approved immediately), and thus the pre-checking which rejection we are handling here is the sole owner of the `code_hash`

[^1]: The condition is actually more complex: it's no such code or pvf pre-checking disabled. If it is disabled though we would not reject the PVF.